tiledb/sm/array_schema/dimension.h

```diff
@@ -141,6 +141,172 @@ class Dimension {
+   *
+   * @param v The value.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The value rounded down to the tile boundary.
+   */
+  template <
+      class T,
+      typename std::enable_if<!std::is_integral<T>::value>::type* = nullptr>
+  static T round_to_tile(
+      const T& v, const T& domain_low, const T& tile_extent) {
+    return (v - domain_low) / tile_extent * tile_extent + domain_low;
+  }
+
+  /**
+   * Returns the tile lower coordinate for integer values.
+   *
+   * @param tile_num The tile index.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The tile lower coordinate.
+   */
+  template <
+      class T,
+      typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  static T tile_coord_low(
+      uint64_t tile_num, const T& domain_low, const T& tile_extent) {
+    typedef typename std::make_unsigned<T>::type unsigned_t;
+    return (unsigned_t)domain_low + tile_num * (unsigned_t)tile_extent;
+  }
+
+  /**
+   * Returns the tile lower coordinate for floating point values.
+   *
+   * @param tile_num The tile index.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The tile lower coordinate.
+   */
+  template <
+      class T,
+      typename std::enable_if<!std::is_integral<T>::value>::type* = nullptr>
+  static T tile_coord_low(
+      uint64_t tile_num, const T& domain_low, const T& tile_extent) {
+    return domain_low + tile_num * tile_extent;
+  }
+
+  /**
+   * Returns the tile upper coordinate for integer values.
+   *
+   * @param tile_num The tile index.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The tile upper coordinate.
+   */
+  template <
+      class T,
+      typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  static T tile_coord_high(
+      uint64_t tile_num, const T& domain_low, const T& tile_extent) {
+    typedef typename std::make_unsigned<T>::type unsigned_t;
+    if ((unsigned_t)tile_extent == std::numeric_limits<unsigned_t>::max())
+      return std::numeric_limits<T>::max() -
+             (domain_low == std::numeric_limits<T>::min());
+    return (unsigned_t)domain_low + ++tile_num * (unsigned_t)tile_extent - 1;
+  }
+
+  /**
+   * Returns the tile upper coordinate for floating point values.
+   *
+   * @param tile_num The tile index.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The tile upper coordinate.
+   */
+  template <
+      class T,
+      typename std::enable_if<!std::is_integral<T>::value>::type* = nullptr>
+  static T tile_coord_high(
+      uint64_t tile_num, const T& domain_low, const T& tile_extent) {
+    return domain_low + ++tile_num * tile_extent - 1;
```

Comment:
I think this should be `return domain_low + ++tile_num * tile_extent - std::numeric_limits::epsilon( );`

Reply:
std::epsilon() wouldn't be reliable here, since it's scaled to 1.0 in its magnitude. If the domain limits are much larger than that, the subtraction will have no effect.
The larger issue that we don't have a specification for whether the bounds represent a closed interval, a closed interval, a half-closed interval, etc. Being precise about these intervals allows a single number to represent a common boundary point without unreliable numeric tricks. The lower interval would be upper-open (resp. upper-closed) and the upper interval would lower-closed (resp. lower open). The common point, in other words, is part of exactly one of the intervals.

Reply:
Then we can probably use `std::next_after` for `domain_low + ++tile_num * tile_extent`.

Currently all ranges for us are closed. Of course we'd like to relax that in the near future.